src/tracks/playabletrack/wavetrack/ui/ChangeClipSpeedDialog.cpp

```diff
@@ -0,0 +1,156 @@
+         parent, wxID_ANY, XO("Change Speed"), wxDefaultPosition, { 480, 250 },
+         wxDEFAULT_DIALOG_STYLE)
+    , mTrack { waveTrack }
+    , mTrackInterval { interval }
+    , mClipSpeed { 100.0 / interval.GetStretchRatio() }
+    , mOldClipSpeed { mClipSpeed }
+{
+   ShuttleGui s(this, eIsCreating);
+
+   s.StartVerticalLay();
+   {
+      PopulateOrExchange(s);
+   }
+   s.EndVerticalLay();
+
+   // TODO: Tolerance?
+   // Stretch ratio of 
+   assert(mOldClipSpeed > 0.0);
+
+   Layout();
+   Fit();
+   Centre();
+
+   Bind(
+      wxEVT_CHAR_HOOK,
+      [this](auto& evt)
+      {
+         if (!IsEscapeKey(evt))
+         {
+            evt.Skip();
+            return;
+         }
+
+         EndModal(wxID_CANCEL);
+      });
+}
+
+ChangeClipSpeedDialog::~ChangeClipSpeedDialog() = default;
+
+void ChangeClipSpeedDialog::PopulateOrExchange(ShuttleGui& s)
+{
+   s.StartInvisiblePanel(15);
+   s.StartVerticalLay(wxEXPAND, 1);
+   {
+      s.SetBorder(1);
+      s.StartHorizontalLay(wxEXPAND);
+      {
+         s.AddPrompt(XO("Clip speed"));
+         s.AddSpace(0, 0, 1);
+         s.Name(XO("Clip speed"))
+            .NameSuffix(Verbatim("%"))
+            .TieNumericTextBox({}, mClipSpeed, 14, true)
+            ->Bind(wxEVT_TEXT_ENTER, [this](auto&) { OnOk(); });
+
+         s.AddFixedText(Verbatim("%"));
+      }
+      s.EndMultiColumn();
+   }
+   s.EndHorizontalLay();
+   s.EndInvisiblePanel();
+
+   s.StartHorizontalLay(wxEXPAND, 0);
+   {
+      s.StartInvisiblePanel(14);
+      {
+         s.SetBorder(2);
+         s.StartHorizontalLay(wxEXPAND, 0);
+         {
+            s.AddSpace(0, 0, 1);
+
+            s.AddButton(XXO("&Cancel"))
+               ->Bind(wxEVT_BUTTON, [this](auto&) { EndModal(wxID_CANCEL); });
+
+            s.AddSpace(4, 0, 0);
+
+            auto okBtn = s.AddButton(XXO("&Ok"));
+            okBtn->Bind(wxEVT_BUTTON, [this](auto&) { OnOk(); });
+            okBtn->SetDefault();
+         }
+         s.EndHorizontalLay();
+      }
+      s.EndInvisiblePanel();
+   }
+   s.EndHorizontalLay();
+}
+
+void ChangeClipSpeedDialog::OnOk()
+{
+   if (ApplyStretchRatio())
+      EndModal(wxID_OK);
+}
+
+bool ChangeClipSpeedDialog::ApplyStretchRatio()
+{
+   {
+      ShuttleGui S(this, eIsGettingFromDialog);
+      PopulateOrExchange(S);
+   }
+
+   if (mClipSpeed <= 0.0)
+   {
+      BasicUI::ShowErrorDialog(
+         /* i18n-hint: Title of an error message shown, when invalid clip speed is set */
+         wxWidgetsWindowPlacement { this }, XO("Invalid clip speed"),
+         XO("Clip speed must be a positive value"), {});
+
+      return false;
+   }
+
+   const auto nextClip = mTrack.GetNextInterval(mTrackInterval, PlaybackDirection::forward);
+   const auto maxEndTime = nextClip != nullptr ?
+                              nextClip->Start() :
+                              std::numeric_limits<double>::infinity();
+
+   const auto expectedEndTime = mTrackInterval.End() * mOldClipSpeed / mClipSpeed;
```

Comment:
I think this is the reason for the problem I found. It should be something like `interval.Begin() + (interval.End()-interval.Begin())*mOldSpeed/mSpeed;`

Reply:
Yes, you are totally right